### PR TITLE
rabbitmq_queue, add param to allow HTTPS connection to mgt api, fixes #22953

### LINKED ISSUE #22953: rabbitmq_queue has no option to use HTTPS when accessing RabbitMQ management API
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
rabbitmq_queue

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 6cea89299a) last updated 2017/03/24 14:57:26 (GMT -400)
  config file = /home/redacted/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### CONFIGURATION
< destination_variable = private_dns_name
< vpc_destination_variable = private_ip_address

##### OS / ENVIRONMENT
Controller:
```
> cat /etc/lsb-release 
DISTRIB_ID=LinuxMint
DISTRIB_RELEASE=18
DISTRIB_CODENAME=sarah
DISTRIB_DESCRIPTION="Linux Mint 18 Sarah"
```
Target:
```
> cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.5 LTS"
```

##### SUMMARY
The module hardcodes "http" in the URI which it generates which prevents it from accessing systems which are expecting HTTPS
`url = "http://%s:%s/api/queues/%s/%s" % (`

##### STEPS TO REPRODUCE
Use the module to create a queue on a RabbitMQ instance which requires HTTPS for the management interface.

```yaml
- name: Create queues
  rabbitmq_queue:
    name: testq
    vhost: "{{ rmq_vhost }}"
    login_user: "{{ rmq_user }}"
    login_password: "{{ rmq_pass }}"
    login_host: "{{ ansible_fqdn }}"
    login_port: 15671
    durable: True
```
##### EXPECTED RESULTS
The queue 'testq' should be created under vhost "{{ rmq_vhost }}"'  on the RabbitMQ instance running at  "{{ ansible_fqdn }}" who's management interface is using HTTPS on port 15671

##### ACTUAL RESULTS

The API call fails because it's using HTTP instead of HTTPS

```
Using module file /home/redacted/Ansible/ansible/lib/ansible/modules/messaging/rabbitmq_queue.py
<172.16.191.87> ESTABLISH SSH CONNECTION FOR USER: automation
<172.16.191.87> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=automation -o ConnectTimeout=10 -o ControlPath=/home/redacted/.ansible/cp/999765f0fe 172.16.191.87 '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-cfjgpsspmnwpsvntxvjablneyhyiyfvr; redacted/pyenvs/ansible/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<172.16.191.87> (1, '', 'Traceback (most recent call last):\n  File "/tmp/ansible_qDVLbU/ansible_module_rabbitmq_queue.py", line 277, in <module>\n    main()\n  File "/tmp/ansible_qDVLbU/ansible_module_rabbitmq_queue.py", line 167, in main\n    r = requests.get( url, auth=(module.params[\'login_user\'],module.params[\'login_password\']))\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/api.py", line 70, in get\n    return request(\'get\', url, params=params, **kwargs)\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/api.py", line 56, in request\n    return session.request(method=method, url=url, **kwargs)\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/sessions.py", line 488, in request\n    resp = self.send(prep, **send_kwargs)\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/sessions.py", line 609, in send\n    r = adapter.send(request, **kwargs)\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/adapters.py", line 487, in send\n    raise ConnectionError(e, request=request)\nrequests.exceptions.ConnectionError: HTTPConnectionPool(host=\'testhost.some.domain\', port=15671): Max retries exceeded with url: /api/queues/%2Fhec%2Ftest/recording.upload.processing (Caused by NewConnectionError(\'<requests.packages.urllib3.connection.HTTPConnection object at 0x7fca0cc12310>: Failed to establish a new connection: [Errno 111] Connection refused\',))\n')
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_qDVLbU/ansible_module_rabbitmq_queue.py", line 277, in <module>
    main()
  File "/tmp/ansible_qDVLbU/ansible_module_rabbitmq_queue.py", line 167, in main
    r = requests.get( url, auth=(module.params['login_user'],module.params['login_password']))
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/api.py", line 70, in get
    return request('get', url, params=params, **kwargs)
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/api.py", line 56, in request
    return session.request(method=method, url=url, **kwargs)
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/sessions.py", line 488, in request
    resp = self.send(prep, **send_kwargs)
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/sessions.py", line 609, in send
    r = adapter.send(request, **kwargs)
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/adapters.py", line 487, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='testhost.some.domain', port=15671): Max retries exceeded with url: /api/queues/%2Ffubar/testq (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x7fca0cc12310>: Failed to establish a new connection: [Errno 111] Connection refused',))

```

### PR DESCRIPTION
**Pull request by allpoints**
##### SUMMARY
Adds parameter to rabbitmq_queue to allow it to make HTTPS requests against the management API. 

New parameter: 
```
    login_with_ssl:
        description:
            - use SSL when accessing RabbitMQ host
        required: false
        default: false
```
If the parameter is not passed or is False, there  is no change from the previous behavior.  If the parameter is set to True, the URL is set to "https://...."

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
rabbitmq_queue

##### ANSIBLE VERSION

```
ansible 2.4.0 (devel 6cea89299a) last updated 2017/03/24 14:57:26 (GMT -400)
  config file = /home/redacted/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

Output after change:
```
Using module file /home/predacted/Ansible/ansible/lib/ansible/modules/messaging/rabbitmq_queue.py
<172.16.191.87> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAu
<172.16.191.87> (0, '\n{"invocation": {"module_args": {"login_host": "localhost.some.domain", "login_port": "15671", "dead_let
ok: [172.16.191.87] => (item=testq) => {
    "changed": false,
    "invocation": {
        "module_args": {
            "arguments": {},
            "auto_delete": false,
            "auto_expires": null,
            "dead_letter_exchange": null,
            "dead_letter_routing_key": null,
            "durable": true,
            "login_host": "localhost.some.domain",
            "login_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "login_port": "15671",
            "login_user": "web_api",
            "login_with_ssl": true,
            "max_length": null,
            "message_ttl": null,
            "name": "testq",
            "state": "present",
            "vhost": "/fubar"
        }
    },
    "item": "testq",
    "name": "testq"

```

### CODE REVIEW

**Comment by manuel-sousa** [comment]
Thanks @allpoints, could you please fix the error above by adding version_added on the new option as well as make the change on rabbitmq_exchange, rabbitmq_binding as both of them use the same interface?

**Comment by allpoints** [comment]
Sure.  I'll take care of it this afternoon.

**Comment by manuel-sousa** [comment]
shipit

File: lib/ansible/modules/messaging/rabbitmq_binding.py [hunk 1]
```diff
@@ -67,6 +67,12 @@
             - rabbitMQ management api port
         required: false
         default: 15672
+    login_with_ssl:
```
Comment (bcoca) [inline lib/ansible/modules/messaging/rabbitmq_binding.py:70]: just to bikeshed use_ssl or use_tls?

**Comment by bcoca** [comment]
one note, you can use document fragments and centralize common fields instead of having to copy the same in each module.

File: lib/ansible/modules/messaging/rabbitmq_binding.py (hunk 1 above)
Comment (flowerysong) [reply lib/ansible/modules/messaging/rabbitmq_binding.py:70]: login_ matches the other parameters that define the host connection, but I think login_tls would be a better name for this one.

**Comment by bcoca** [comment]
As for the argspec, if you have common fields, these can also bet setup in module_utils/ for all modules to reuse.

File: lib/ansible/modules/messaging/rabbitmq_binding.py (hunk 1 above)
Comment (allpoints) [reply lib/ansible/modules/messaging/rabbitmq_binding.py:70]: Agreed, happy to change it.

**Comment by Jorge-Rodriguez** [comment]
This PR enables the HTTPS protocol only, but it does not utilize Requests' TLS functionalities. If `login_with_ssl == True` the request will use Requests' bundled certifi certificates or whatever bundle is set on the REQUESTS_CA_BUNDLE environment variable. Also, there is no entrypoint to provide client certificates.

See #18437

**Comment by Akasurde** [comment]
@allpoints Could you please rebase your changes ? needs_info

**Comment by Jorge-Rodriguez** [comment]
@Akasurde why are you asking for this rebase when there's another PR that supersedes this one and is being maintained and waiting for merge for close to three years?
Please check #18437 and see if it gets merged before hell freezes over.

**Comment by Akasurde** [comment]
Superseded by #18437

**Pull request closed without merging**